File: offer/26.树的子结构.js
function isSubStructure(A, B) {
  if (!A || !B) {
    return false;
  }
  return isSubTree(A, B) || isSubStructure(A.left, B) || isSubStructure(A.right, B);
}

function isSubTree(A, B) {
  if (!B) {
    return true;
  }
  if (!A || A.val !== B.val) {
    return false;
  }
  return isSubTree(A.left, B.left) && isSubTree(A.right, B.right);
}

File: offer/26.树的子结构.go
func isSubStructure(A *TreeNode, B *TreeNode) bool {
	if A == nil || B == nil {
		return false
	}
	return isSubTree(A, B) || isSubStructure(A.Left, B) || isSubStructure(A.Right, B)
}

func isSubTree(A *TreeNode, B *TreeNode) bool {
	if B == nil {
		return true
	}
	if A == nil || A.Val != B.Val {
		return false
	}
	return isSubTree(A.Left, B.Left) && isSubTree(A.Right, B.Right)
}